text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,335 @@
+
+For context, the current design requires a component author define an arbitrary custom element name which will be used as the name for all constructors passed to the `lwc:dynamic` directive, creating a 1:N mapping.
+
+For example:
+
+```html
+<template>
+  <x-lazy lwc:dynamic={customCtor}></x-lazy>
+</template>
+```
+
+```javascript
+import { LightningElement } from "lwc";
+
+export default class DynamicCtor extends LightningElement {
+  customCtor;
+
+  connectedCallback() {
+    this.loadCtor();
+  }
+
+  async loadCtor() {
+    const { default: ctor } = await import("c/customConstructor");
+    this.customCtor = ctor;
+  }
+}
+```
+
+Will always render the same tag regardless of the constructor.
+
+```html
+  <x-lazy>
+    <!-- c-custom-constructor content -->
+  </x-lazy>
+```
+
+This proposal aims to overcome the issues with the current design by exposing the custom element name at compile time and using it as the name of the dynamic component that is created at runtime.
+
+## Prior art
+
+Several other frameworks have the concept of dynamically rendering components:
+
+- [Svelte: `<svelte:component this={expression}>`](https://svelte.dev/docs#template-syntax-svelte-component)
+- [Vue: `<component :is="...">`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
+- [Angular: viewContainerRef](https://angular.io/guide/dynamic-component-loader)
+- [Lit: static expressions](https://lit.dev/docs/templates/expressions/#static-expressions)
+- [Aura: $A.createComponent](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_cmp_async.htm)
+
+Frameworks that use JSX such as [React](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom) and [Stencil](https://stenciljs.com/docs/templating-jsx) can dynamically render components directly into the markup using expressions.
+
+The goal of managing dynamic components in LWC is to follow the same component lifecycle as all other LWC custom elements.  Having said that, taking a similar approach as Svelte and Vue seems the most natural and intuitive path forward.
+
+## Detailed design
+
+A new property, `lightningElementName` on `LightningElementConstructor` is introduced to store the name of the custom element.
+
+A special tag, `<lwc.dynamic>` along with a new template directive `lwc:is` are also introduced and serves as the anchor in the DOM where the dynamic component will be rendered.
+
+Broken down, there are three main parts to this design:
+
+1. Storing and retrieving a custom element's name.
+2. Defining an anchor in the DOM.
+3. Instantiating the dynamic custom element.
+
+#### Storing and retrieving a custom element's name
+
+In today's world, an LWC module's namespace and name are known at compile time by [examining the directory structure](https://rfcs.lwc.dev/rfcs/lwc/0101-module-unification) and through [module resolution](https://rfcs.lwc.dev/rfcs/lwc/0020-module-resolution).
```

Comment:
Side note: I am suspecting that we might have to do some work in the `@lwc/rollup-plugin`. I know that in the past the `<namespace>/<name>/<name>.js` folder structure was not always respected in off-core projects. It's something that we might have fixed over the years.

https://github.com/salesforce/lwc/blob/3fafcac1ae4b1887c1b5213d076d02d559dd7549/packages/%40lwc/rollup-plugin/src/index.ts#L214-L215

Reply:
Thanks for calling this out!  I was wondering if there may be some cases when the folder structure mapping would be off. 

I'll make a note to monitor this and do some research to see what the current state of affairs is.

Reply:
@pmdartus just to clarify, if the folder structure does not respect the standard `<namespace>/<name>/<name>.js` format the only way it can be recognized is if it has an [alias module record](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/module-resolver#alias-module-record) correct?

Reply:
> @pmdartus just to clarify, if the folder structure does not respect the standard `<namespace>/<name>/<name>.js` format the only way it can be recognized is if it has an [alias module record](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/module-resolver#alias-module-record) correct?

Correct.